src/main/java/bookrepo/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,107 @@
+    private final CartItemRepository cartItemRepository;
+
+    @Override
+    public ShoppingCartDto getShoppingCart() {
+        User user = authenticationService.getAuthenticatedUser();
+        ShoppingCart cart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart not found for user id: "
+                        + user.getId()));
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    public ShoppingCartDto save(CreateCartItemDto requestDto) {
+        User user = authenticationService.getAuthenticatedUser();
+
+        ShoppingCart cart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> {
+                    createShoppingCartForUser(user);
+                    return shoppingCartRepository.findByUserId(user.getId())
+                            .orElseThrow(() -> new EntityNotFoundException(
+                                    "Shopping cart not found for user id: "
+                                    + user.getId()));
+                });
+
+        Book book = bookRepository.findById(requestDto.getBookId())
+                .orElseThrow(() -> new EntityNotFoundException("Book: "
+                        + requestDto.getBookId()
+                        + " not found"));
+
+        Optional<CartItem> existingItem = cart.getCartItems().stream()
+                .filter(item -> item.getBook().getId().equals(book.getId()))
+                .findFirst();
+
+        if (existingItem.isPresent()) {
+            existingItem.get().setQuantity(existingItem.get().getQuantity()
+                    + requestDto.getQuantity());
+        } else {
+            CartItem newItem = new CartItem();
+            newItem.setBook(book);
+            newItem.setQuantity(requestDto.getQuantity());
+            newItem.setShoppingCart(cart);
+            cart.getCartItems().add(newItem);
+        }
+
+        shoppingCartRepository.save(cart);
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    public ShoppingCartDto update(Long cartItemId, UpdateCarItemQuantityDto quantity) {
+        User user = authenticationService.getAuthenticatedUser();
+
+        ShoppingCart cart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Shopping cart not found"));
+
+        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Cart item not found"));
+
+        cartItem.setQuantity(quantity.getQuantity());
+
+        shoppingCartRepository.save(cart);
+
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    public void deleteById(Long cartItemId) {
+        cartItemRepository.deleteById(cartItemId);
+    }
+
+    private void createShoppingCartForUser(User user) {
```

Comment:
add method in interface
```suggestion
    @Override
    public void createShoppingCartForUser(User user) {
```